PsfLauncher/main.cpp

```diff
@@ -13,223 +13,227 @@
+        }
     }
-    else
+
+    //If we get here we know that starting script did NOT encounter an error
+    //run the monitor.
+    //Launch monitor if we are using one.
+    auto monitor = PSFQueryAppMonitorConfig();
+    if (monitor != nullptr)
     {
-        // Use requested path, relative to the package root folder.
-        std::wstring wdwd = (packageRoot / dirStr).native();
-        wdwd.resize(wdwd.size() - 1); // remove trailing slash
-        wd =  wdwd ;
+        error = GetAndLaunchMonitor(monitor, packageRoot, cmdShow, dirStr);
     }
-    std::wstring quotedapp = exePath.native(); // L"\"" + exePath.native() + L"\"";
 
-    if (check_suffix_if(exeName, L".exe"_isv))
+    if (!error.IsThereAnError())
     {
-        STARTUPINFO startupInfo = { sizeof(startupInfo) };
-        startupInfo.dwFlags = STARTF_USESHOWWINDOW;
-        startupInfo.wShowWindow = static_cast<WORD>(cmdShow);
-
-        Log("\tCreating process %ls", cmdLine.data());
-        // NOTE: PsfRuntime, which we dynamically link against, detours this call
-        PROCESS_INFORMATION processInfo;
-        if (!::CreateProcessW(
-            exePath.c_str(),
-            cmdLine.data(),
-            nullptr, nullptr, // Process/ThreadAttributes
-            true, // InheritHandles
-            0, // CreationFlags
-            nullptr, // Environment
-            dirStr ? (packageRoot / dirStr).c_str() : nullptr, // NOTE: extended-length path not supported
-            &startupInfo,
-            &processInfo))
+        //Launch underlying application.
+        auto exeName = appConfig->get("executable").as_string().wide();
+        auto exePath = packageRoot / exeName;
+        auto exeArgString = exeArgs ? exeArgs->as_string().wide() : (wchar_t*)L"";
+        // Allow arguments to be specified in config.json now also.
+        std::wstring cmdLine = L"\"" + exePath.filename().native() + L"\" " + exeArgString + L" " + args;
+        if (check_suffix_if(exeName, L".exe"_isv))
         {
-            std::wostringstream ss;
-            auto err{ ::GetLastError() };
-            // Remove the ".\r\n" that gets added to all messages
-            auto msg = widen(std::system_category().message(err));
-            msg.resize(msg.length() - 3);
-            ss << L"ERROR: Failed to create detoured process\n  Path: \"" << exeName << "\"\n  Error: " << msg << " (" << err << ")";
-            ::PSFReportError(ss.str().c_str());
-            return err;
-        }
+            std::wstring workingDirectory;
 
-        // Propagate exit code to caller, in case they care
-        switch (::WaitForSingleObject(processInfo.hProcess, INFINITE))
-        {
-        case WAIT_OBJECT_0:
-            break; // Success case... handle at the end of main
+            if (dirStr)
+            {
+                workingDirectory = (packageRoot / dirStr).native();
+            }
 
-        case WAIT_FAILED:
-            return ::GetLastError();
+            ExecutionInformation execInfo;

```

Comment:
consider just using aggregate init:  execInfo{ ... }

Reply:
Sure.  I'm not used to doing an aggregated init for my own objects, but it does look nicer.

Do you know if making an aggregated init is faster?

Reply:
Wait.  No.  I'm confused now.  Are you suggesting to write the code like this

`ExecutionInformation execInfo {exePath.c_str(), cmdLine}` instead?

Reply:
nicer and faster (as you're avoiding double-initialization, which is the same reason you should prefer constructor init-lists)